### PR TITLE
Introduce minReadySeconds in StatefulSets

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
Introduce minReadySeconds in StatefulSets

### CODE REVIEW

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+with other workload controllers.
+
+### Goals
+
+StatefulSet Controller honoring minReadySeconds and mark Pod ready only if Pod 
+is available for given time mentioned in minReadySeconds.
+
+### Non-Goals
+
+Moving minReadySeconds to Pod Spec is beyond the scope of this KEP
```
Comment (smarterclayton) [inline keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:174]: I would describe why, and maybe say a future story could address it based on consistent use.  First, the effort to change pod spec would be large (impact).  Second, currently our workload controllers are inconsistent and we prioritize consistent of experience over desirable future state (consistency).  Third, stateful sets are just different enough from daemonsets and deployments that real world use of minReadySeconds for stateful sets might influence any future design or point in a more appropriate direction (completeness)

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+ `minReadySeconds`, there are two cases
+  - If `minReadySeconds` is greater than 0 -- in this case kube-apiserver
+    clears the `minReadySeconds` and the existing StatefulSets wouldn't honor 
+    `minReadySeconds` which is expected. The same is the case with `AvailableReplicas`
+  - If `minReadySeconds` is equal to 0 -- in this case user wont see any 
+    difference in behavior
+
+### Version Skew Strategy
+
+This feature has no node runtime implications
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:310]: > What happens when you have a statefulset deployed with a non-zero minReadySeconds on a cluster with kubelets on version n-2 or n-1?

Kubelet is not effected by introduction of this api.

> What about a controller-manager on version n-1 with an API server on version n?

If the feature gate is enabled, the controller manager updates the `AvailableReplicas` to reflect the pods that have been available for `minReadySeconds`as mentioned, if not the controller manager runs as is.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+    - kube-controller-manager
+    - kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+No, StatefulSet needs to have `.spec.minReadySeconds` explicitly set
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Using the featuregate is the recommended way
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:356]: Nope. As mentioned earlier, if the field on StatefulSet spec is unset or set to 0, `AvailableReplicas` will be updated immediately to `ReadyReplicas` if the featuregate is enabled.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The StatefulSet controller starts respecting the `minReadySeconds` again. 
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes, unit and integration tests for feature enabled, disabled and transitioned
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:364]: Which part can we not do using integration tests? I thought we can enable and disable featuregates using:

```go
 // run with gate on
   defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features. Frobber2D, true)()
// run with gate off, *do not assume it is off by default*
   defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features. Frobber2D, false)()
 ```

**Review by ravisantoshgudimetla — COMMENTED** [review]
@ehashman - Thank you for the review. I tried answering your questions, can you PTAL

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+service loadbalancers in clouds. This is to prevent killing pods in rotation
+before new pods show up.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+We are proposing a new field called `minReadySeconds` whose default value will 
+be 0. In this mode, StatefulSet will behave exactly like its current behavior.
+Its possible we introduce a bug in the implementation. The mitigation
```
Comment (soltysh) [inline keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:207]: Be more specific about the risk of a bug, namely that we either:
- mark a pod as ready too soon
- mark as ready too late. 
and what that might potentially mean.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+### StatefulSet
+
+```go
+type StatefulSetSpec struct {
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	// This is an alpha field and requires enabling StatefulSetMinReadySeconds feature gate.
+	// +optional
+	MinReadySeconds *int32
```
Comment (soltysh) [inline keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:225]: For consistency with other controllers this should not be a pointer, especially that 0 and nil have the same meaning.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+}
+
+type StatefulSetStatus struct {
+	// Total number of available pods (ready for at least minReadySeconds) targeted by this statefulset.
+	// This is an alpha field and requires enabling StatefulSetMinReadySeconds feature gate.
+	// +optional
+	AvailabileReplicas *int32
```
Comment (soltysh) [inline keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:232]: Ditto.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+**For non-optional features moving to GA, the graduation criteria must include 
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+-->
+
+### Upgrade / Downgrade Strategy
+- Upgrades
+ When upgrading from a release without this feature, to a release with 
+ `minReadySeconds`, we will set `minReadySeconds` to 0. This would give users
```
[the remark this answers is not in the document]
Comment (soltysh) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:296]: See my other comment about this not being a pointer.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md (hunk 6 above)
Comment (ravisantoshgudimetla) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:225]: Ya, I was thinking about the same but in the [apiconventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md), it was mentioned that:


- ensure the field is entirely absent from API responses when empty (**optional fields should be pointers, anyway**) which made me think I should use a pointer. It also helps in setting default values easily as we can  distinguish between  being set and the default value.

This was also mentioned in the api conventions doc.
```
Do use pointers to scalars when you need to distinguish between an unset value and an automatic zero value. For example, PodSpec.TerminationGracePeriodSeconds is defined as *int64 the go type definition. A zero value means 0 seconds, and a nil value asks the system to pick a default.
```

I think the above is the case for us too say if we choose to a non-zero default value in future.

Comment (soltysh) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:225]: We spoke with @smarterclayton on slack, and the agreement is that we'll go with non-pointer version.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+ the same default behavior. The default value of `AvailableReplicas` would
+ be value of existing field in Status sub-resource called `ReadyReplicas`
+- Downgrades
+ When downgrading from a release with this feature, to a release without 
+ `minReadySeconds`, there are two cases
+  - If `minReadySeconds` is greater than 0 -- in this case kube-apiserver
+    clears the `minReadySeconds` and the existing StatefulSets wouldn't honor 
+    `minReadySeconds` which is expected. The same is the case with `AvailableReplicas`
+  - If `minReadySeconds` is equal to 0 -- in this case user wont see any 
+    difference in behavior
```
[the remark this answers is not in the document]
Comment (ehashman) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:318]: Please update the KEP to include this design detail; I think this is core design, not just an implementation detail.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehashman) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:322]: Please update the KEP to say that.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ehashman) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:368]: Please ensure this is well-documented.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ehashman) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:376]: You can turn the feature gates on and off at the beginning of E2Es, but afaik it is not possible to disable a feature flag midway through an e2e when it started enabled, so this behaviour should be manually validated.

File: keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md:318]: Done. Addressed it in the below paragraph

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Review by ehashman — APPROVED** [review]
/approve
for PRR

**Comment by chrisnegus** [comment]
Hi @whomever 👋 1.22 Docs Shadow here.
This enhancement is marked as Needs Docs for the 1.22 release.

Please follow the steps detailed in the[ documentation to open a PR](https://kubernetes.io/docs/contribute/new-content/new-features/#open-a-placeholder-pr) against dev-1.22 in the kubernetes/website repo. This PR can be just a placeholder at this time and must be created before *Fri July 9, 11:59 PM PDT*.

Also, take a look at [Documenting for a release](https://kubernetes.io/docs/contribute/new-content/new-features/#for-developers-or-other-sig-members) to familiarize yourself with the docs requirement for the release.

Thank you!

**Pull request merged**